File: frontend/src/types/actions.ts
/**
 * Action types for checkpoint-based validation pipelines.
 *
 * Actions are executed after validation completes. They can store results,
 * send notifications, or integrate with external systems.
 *
 * Supported action types:
 * - Storage: Store validation results to filesystem, S3, GCS
 * - Notifications: Slack, Email, Teams, Discord, Telegram, PagerDuty, OpsGenie
 * - Webhook: Call any HTTP endpoint
 * - Custom: Execute Python callbacks or shell commands
 * - DataDocs: Generate HTML/Markdown documentation
 */

// =============================================================================
// Action Status and Conditions
// =============================================================================

/**
 * Status of an action execution.
 */
export type ActionStatus =
  | 'pending'      // Not yet started
  | 'running'      // Currently executing
  | 'success'      // Completed successfully
  | 'failure'      // Failed
  | 'skipped'      // Skipped due to notify_on condition
  | 'compensated'  // Rolled back (for transactional actions)

/**
 * Conditions under which actions are triggered.
 */
export type NotifyCondition =
  | 'always'           // Every run
  | 'success'          // Validation passed
  | 'failure'          // Validation failed
  | 'error'            // System error occurred
  | 'failure_or_error' // Failure or error
  | 'not_success'      // Any non-success status
  | 'warning'          // Warning status

/**
 * Human-readable labels for notify conditions.
 */
export const NOTIFY_CONDITION_LABELS: Record<NotifyCondition, string> = {
  always: 'Always',
  success: 'On Success',
  failure: 'On Failure',
  error: 'On Error',
  failure_or_error: 'On Failure or Error',
  not_success: 'Not Success',
  warning: 'On Warning',
}

// =============================================================================
// Action Types
// =============================================================================

/**
 * Available action types.
 */
export type ActionType =
  | 'slack'
  | 'email'
  | 'teams'
  | 'discord'
  | 'telegram'
  | 'pagerduty'
  | 'opsgenie'
  | 'webhook'
  | 'store_result'
  | 'update_docs'
  | 'custom'
  | 'callback'
  | 'shell_command'
  | 'file_storage'
  | 's3_storage'
  | 'gcs_storage'

/**
 * Action type categories.
 */
export type ActionCategory = 'notification' | 'storage' | 'integration' | 'custom'

/**
 * Action type metadata.
 */
export interface ActionTypeInfo {
  type: ActionType
  label: string
  description: string
  category: ActionCategory
  icon?: string
  configSchema?: ActionConfigSchema
}

/**
 * Action config field schema.
 */
export interface ActionConfigField {
  name: string
  type: 'string' | 'number' | 'boolean' | 'select' | 'array' | 'object'
  label: string
  description?: string
  required?: boolean
  default?: unknown
  options?: { value: string; label: string }[]
  validation?: {
    min?: number
    max?: number
    pattern?: string
    message?: string
  }
}

/**
 * Schema for action configuration.
 */
export interface ActionConfigSchema {
  fields: ActionConfigField[]
}

/**
 * Metadata for all action types.
 */
export const ACTION_TYPE_INFO: Record<ActionType, ActionTypeInfo> = {
  slack: {
    type: 'slack',
    label: 'Slack',
    description: 'Send notifications via Slack webhook',
    category: 'notification',
    icon: 'slack',
  },
  email: {
    type: 'email',
    label: 'Email',
    description: 'Send email notifications',
    category: 'notification',
    icon: 'mail',
  },
  teams: {
    type: 'teams',
    label: 'Microsoft Teams',
    description: 'Send notifications via Teams webhook',
    category: 'notification',
    icon: 'users',
  },
  discord: {
    type: 'discord',
    label: 'Discord',
    description: 'Send notifications via Discord webhook',
    category: 'notification',
    icon: 'message-circle',
  },
  telegram: {
    type: 'telegram',
    label: 'Telegram',
    description: 'Send notifications via Telegram bot',
    category: 'notification',
    icon: 'send',
  },
  pagerduty: {
    type: 'pagerduty',
    label: 'PagerDuty',
    description: 'Create PagerDuty incidents',
    category: 'notification',
    icon: 'alert-triangle',
  },
  opsgenie: {
    type: 'opsgenie',
    label: 'OpsGenie',
    description: 'Create OpsGenie alerts',
    category: 'notification',
    icon: 'bell',
  },
  webhook: {
    type: 'webhook',
    label: 'Webhook',
    description: 'Call any HTTP endpoint',
    category: 'integration',
    icon: 'globe',
  },
  store_result: {
    type: 'store_result',
    label: 'Store Result',
    description: 'Store validation results to storage',
    category: 'storage',
    icon: 'database',
  },
  update_docs: {
    type: 'update_docs',
    label: 'Update DataDocs',
    description: 'Generate HTML/Markdown documentation',
    category: 'storage',
    icon: 'file-text',
  },
  custom: {
    type: 'custom',
    label: 'Custom',
    description: 'Execute custom Python code',
    category: 'custom',
    icon: 'code',
  },
  callback: {
    type: 'callback',
    label: 'Callback',
    description: 'Execute a Python callback function',
    category: 'custom',
    icon: 'terminal',
  },
  shell_command: {
    type: 'shell_command',
    label: 'Shell Command',
    description: 'Execute a shell command',
    category: 'custom',
    icon: 'terminal',
  },
  file_storage: {
    type: 'file_storage',
    label: 'File Storage',
    description: 'Store results to local filesystem',
    category: 'storage',
    icon: 'folder',
  },
  s3_storage: {
    type: 's3_storage',
    label: 'S3 Storage',
    description: 'Store results to Amazon S3',
    category: 'storage',
    icon: 'cloud',
  },
  gcs_storage: {
    type: 'gcs_storage',
    label: 'GCS Storage',
    description: 'Store results to Google Cloud Storage',
    category: 'storage',
    icon: 'cloud',
  },
}

// =============================================================================
// Action Configuration
// =============================================================================

/**
 * Base configuration for all actions.
 */
export interface ActionConfigBase {
  /** Action name for identification */
  name: string
  /** Action type */
  type: ActionType
  /** When to execute this action */
  notify_on: NotifyCondition
  /** Whether this action is enabled */
  enabled: boolean
  /** Maximum execution time in seconds */
  timeout_seconds?: number
  /** Number of retries on failure */
  retry_count?: number
  /** Delay between retries in seconds */
  retry_delay_seconds?: number
  /** Fail checkpoint on action error */
  fail_checkpoint_on_error?: boolean
  /** Additional metadata */
  metadata?: Record<string, unknown>
}

/**
 * Slack notification configuration.
 */
export interface SlackActionConfig extends ActionConfigBase {
  type: 'slack'
  /** Slack webhook URL */
  webhook_url: string
  /** Channel override */
  channel?: string
  /** Bot display name */
  username?: string
  /** Bot icon emoji */
  icon_emoji?: string
  /** Include detailed statistics */
  include_details?: boolean
  /** User IDs to mention on failure */
  mention_on_failure?: string[]
  /** Custom message template */
  custom_message?: string
}

/**
 * Email notification configuration.
 */
export interface EmailActionConfig extends ActionConfigBase {
  type: 'email'
  /** SMTP server host */
  smtp_host: string
  /** SMTP server port */
  smtp_port: number
  /** SMTP authentication user */
  smtp_user?: string
  /** SMTP authentication password (env var reference) */
  smtp_password?: string
  /** Use TLS */
  use_tls?: boolean
  /** Use SSL */
  use_ssl?: boolean
  /** Sender address */
  from_address: string
  /** Recipient addresses */
  to_addresses: string[]
  /** CC addresses */
  cc_addresses?: string[]
  /** Subject template */
  subject_template?: string
  /** Include HTML body */
  include_html?: boolean
  /** Provider: smtp, sendgrid, ses */
  provider?: 'smtp' | 'sendgrid' | 'ses'
  /** API key for SendGrid/SES */
  api_key?: string
}

/**
 * Teams notification configuration.
 */
export interface TeamsActionConfig extends ActionConfigBase {
  type: 'teams'
  /** Teams webhook URL */
  webhook_url: string
  /** Channel name for display */
  channel?: string
  /** Include detailed information */
  include_details?: boolean
  /** Message theme */
  theme?: 'auto' | 'success' | 'warning' | 'critical'
}

/**
 * Discord notification configuration.
 */
export interface DiscordActionConfig extends ActionConfigBase {
  type: 'discord'
  /** Discord webhook URL */
  webhook_url: string
  /** Bot display name */
  username?: string
  /** Bot avatar URL */
  avatar_url?: string
  /** Embed color (hex integer) */
  embed_color?: number
  /** Embed title */
  embed_title?: string
  /** Embed description */
  embed_description?: string
  /** Custom embed fields */
  embed_fields?: Array<{ name: string; value: string; inline?: boolean }>
  /** Mentions to include */
  include_mentions?: string[]
}

/**
 * Telegram notification configuration.
 */
export interface TelegramActionConfig extends ActionConfigBase {
  type: 'telegram'
  /** Telegram bot token */
  bot_token: string
  /** Channel/group ID */
  chat_id: string
  /** Parse mode */
  parse_mode?: 'Markdown' | 'HTML'
  /** Custom message template */
  message_template?: string
  /** Silent notification */
  disable_notification?: boolean
  /** Photo URL (optional) */
  photo_url?: string
}

/**
 * PagerDuty notification configuration.
 */
export interface PagerDutyActionConfig extends ActionConfigBase {
  type: 'pagerduty'
  /** PagerDuty routing key */
  routing_key: string
  /** Auto-severity mapping */
  auto_severity?: boolean
  /** Auto-resolve on success */
  auto_resolve_on_success?: boolean
  /** Deduplication key template */
  dedup_key_template?: string
  /** Custom severity override */
  severity_override?: 'critical' | 'error' | 'warning' | 'info'
}

/**
 * OpsGenie notification configuration.
 */
export interface OpsGenieActionConfig extends ActionConfigBase {
  type: 'opsgenie'
  /** OpsGenie API key */
  api_key: string
  /** Alert priority */
  priority?: 'P1' | 'P2' | 'P3' | 'P4' | 'P5'
  /** Tags for the alert */
  tags?: string[]
  /** Responders */
  responders?: Array<{
    type: 'user' | 'team' | 'escalation' | 'schedule'
    name?: string
    username?: string
    id?: string
  }>
  /** Visible to teams */
  visible_to?: Array<{
    type: 'user' | 'team'
    name?: string
    username?: string
    id?: string
  }>
  /** Auto-resolve on success */
  auto_resolve_on_success?: boolean
}

/**
 * Webhook action configuration.
 */
export interface WebhookActionConfig extends ActionConfigBase {
  type: 'webhook'
  /** Webhook URL */
  url: string
  /** HTTP method */
  method?: 'GET' | 'POST' | 'PUT' | 'PATCH'
  /** Authentication type */
  auth_type?: 'none' | 'basic' | 'bearer' | 'api_key'
  /** Auth credentials */
  auth_credentials?: {
    username?: string
    password?: string
    token?: string
    api_key?: string
    api_key_header?: string
  }
  /** Custom headers */
  headers?: Record<string, string>
  /** Include full result in payload */
  include_full_result?: boolean
  /** Request timeout */
  request_timeout?: number
}

/**
 * File storage action configuration.
 */
export interface FileStorageActionConfig extends ActionConfigBase {
  type: 'file_storage' | 'store_result'
  /** Base path for storage */
  store_path: string
  /** Store type */
  store_type?: 'file' | 's3' | 'gcs'
  /** Output format */
  format?: 'json' | 'yaml'
  /** Partition by date, checkpoint, or status */
  partition_by?: 'date' | 'checkpoint' | 'status' | 'none'
  /** Retention days */
  retention_days?: number
  /** Compress output */
  compress?: boolean
}

/**
 * S3 storage action configuration.
 */
export interface S3StorageActionConfig extends ActionConfigBase {
  type: 's3_storage'
  /** S3 bucket */
  bucket: string
  /** Key prefix */
  prefix?: string
  /** AWS region */
  region?: string
  /** Output format */
  format?: 'json' | 'yaml' | 'parquet'
  /** Compression */
  compression?: 'gzip' | 'snappy' | 'none'
}

/**
 * GCS storage action configuration.
 */
export interface GCSStorageActionConfig extends ActionConfigBase {
  type: 'gcs_storage'
  /** GCS bucket */
  bucket: string
  /** Key prefix */
  prefix?: string
  /** GCP project */
  project?: string
  /** Output format */
  format?: 'json' | 'yaml' | 'parquet'
}

/**
 * Update DataDocs action configuration.
 */
export interface UpdateDocsActionConfig extends ActionConfigBase {
  type: 'update_docs'
  /** Site path for documentation */
  site_path: string
  /** Output format */
  format?: 'html' | 'markdown'
  /** Include history */
  include_history?: boolean
  /** Max history items */
  max_history_items?: number
  /** Template name */
  template?: 'default' | 'minimal' | 'detailed'
}

/**
 * Custom action configuration.
 */
export interface CustomActionConfig extends ActionConfigBase {
  type: 'custom' | 'callback'
  /** Python callback function path */
  callback?: string
  /** Environment variables */
  environment?: Record<string, string>
  /** Pass result as JSON */
  pass_result_as_json?: boolean
}

/**
 * Shell command action configuration.
 */
export interface ShellCommandActionConfig extends ActionConfigBase {
  type: 'shell_command'
  /** Shell command to execute */
  shell_command: string
  /** Working directory */
  working_directory?: string
  /** Environment variables */
  environment?: Record<string, string>
  /** Pass result as JSON */
  pass_result_as_json?: boolean
}

/**
 * Union type for all action configurations.
 */
export type ActionConfig =
  | SlackActionConfig
  | EmailActionConfig
  | TeamsActionConfig
  | DiscordActionConfig
  | TelegramActionConfig
  | PagerDutyActionConfig
  | OpsGenieActionConfig
  | WebhookActionConfig
  | FileStorageActionConfig
  | S3StorageActionConfig
  | GCSStorageActionConfig
  | UpdateDocsActionConfig
  | CustomActionConfig
  | ShellCommandActionConfig

// =============================================================================
// Action Result
// =============================================================================

/**
 * Result of an action execution.
 */
export interface ActionResult {
  /** Action name */
  action_name: string
  /** Action type */
  action_type: ActionType
  /** Execution status */
  status: ActionStatus
  /** Human-readable message */
  message?: string
  /** When execution started (ISO string) */
  started_at?: string
  /** When execution completed (ISO string) */
  completed_at?: string
  /** Execution duration in milliseconds */
  duration_ms?: number
  /** Additional result details */
  details?: Record<string, unknown>
  /** Error message if failed */
  error?: string
}

// =============================================================================
// Helper Functions
// =============================================================================

/**
 * Get action type info.
 */
export function getActionTypeInfo(type: ActionType): ActionTypeInfo {
  return ACTION_TYPE_INFO[type]
}

/**
 * Get actions by category.
 */
export function getActionsByCategory(category: ActionCategory): ActionTypeInfo[] {
  return Object.values(ACTION_TYPE_INFO).filter(info => info.category === category)
}

/**
 * Create a default action config for a type.
 */
export function createDefaultActionConfig(type: ActionType, name: string): Partial<ActionConfigBase> {
  return {
    name,
    type,
    notify_on: type === 'store_result' ? 'always' : 'failure',
    enabled: true,
    timeout_seconds: 30,
    retry_count: 0,
    retry_delay_seconds: 5,
    fail_checkpoint_on_error: false,
  }
}

/**
 * Check if action status indicates success.
 */
export function isActionSuccess(status: ActionStatus): boolean {
  return status === 'success'
}

/**
 * Check if action status indicates failure.
 */
export function isActionFailed(status: ActionStatus): boolean {
  return status === 'failure'
}

/**
 * Check if action was skipped.
 */
export function isActionSkipped(status: ActionStatus): boolean {
  return status === 'skipped'
}

File: src/truthound_dashboard/core/interfaces/actions.py
"""Action interfaces for checkpoint-based validation pipelines.

Actions are executed after validation completes. They can store results,
send notifications, or integrate with external systems.

This module defines abstract interfaces for actions that are loosely
coupled from truthound's checkpoint.actions module.

Supported action types:
- Storage: Store validation results to filesystem, S3, GCS
- Notifications: Slack, Email, Teams, Discord, Telegram, PagerDuty
- Webhook: Call any HTTP endpoint
- Custom: Execute Python callbacks or shell commands
- DataDocs: Generate HTML/Markdown documentation
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import TYPE_CHECKING, Any, Callable, Protocol, runtime_checkable

if TYPE_CHECKING:
    from truthound_dashboard.core.interfaces.checkpoint import CheckpointResult


class NotifyCondition(str, Enum):
    """Conditions under which actions are triggered.

    Maps to truthound's notify_on parameter for actions.
    """

    ALWAYS = "always"  # Every run
    SUCCESS = "success"  # Validation passed
    FAILURE = "failure"  # Validation failed
    ERROR = "error"  # System error occurred
    FAILURE_OR_ERROR = "failure_or_error"  # Failure or error
    NOT_SUCCESS = "not_success"  # Any non-success status
    WARNING = "warning"  # Warning status


class ActionStatus(str, Enum):
    """Status of an action execution."""

    PENDING = "pending"
    RUNNING = "running"
    SUCCESS = "success"
    FAILURE = "failure"
    SKIPPED = "skipped"  # Skipped due to notify_on condition
    COMPENSATED = "compensated"  # Rolled back (for transactional actions)


@dataclass
class ActionConfig:
    """Base configuration for all actions.

    Attributes:
        name: Action name for identification.
        notify_on: When to execute this action.
        enabled: Whether this action is enabled.
        timeout_seconds: Maximum execution time.
        retry_count: Number of retries on failure.
        retry_delay_seconds: Delay between retries.
        metadata: Additional metadata for the action.
    """

    name: str = ""
    notify_on: NotifyCondition = NotifyCondition.FAILURE
    enabled: bool = True
    timeout_seconds: int = 30
    retry_count: int = 0
    retry_delay_seconds: int = 5
    metadata: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary."""
        return {
            "name": self.name,
            "notify_on": self.notify_on.value,
            "enabled": self.enabled,
            "timeout_seconds": self.timeout_seconds,
            "retry_count": self.retry_count,
            "retry_delay_seconds": self.retry_delay_seconds,
            "metadata": self.metadata,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "ActionConfig":
        """Create from dictionary."""
        notify_on = data.get("notify_on", "failure")
        if isinstance(notify_on, str):
            notify_on = NotifyCondition(notify_on)
        return cls(
            name=data.get("name", ""),
            notify_on=notify_on,
            enabled=data.get("enabled", True),
            timeout_seconds=data.get("timeout_seconds", 30),
            retry_count=data.get("retry_count", 0),
            retry_delay_seconds=data.get("retry_delay_seconds", 5),
            metadata=data.get("metadata", {}),
        )


@dataclass
class ActionResult:
    """Result of an action execution.

    Attributes:
        action_name: Name of the action.
        action_type: Type of action (notification, storage, etc.).
        status: Execution status.
        message: Human-readable message.
        started_at: When execution started.
        completed_at: When execution completed.
        duration_ms: Execution duration in milliseconds.
        details: Additional result details.
        error: Error message if failed.
    """

    action_name: str
    action_type: str
    status: ActionStatus
    message: str = ""
    started_at: datetime | None = None
    completed_at: datetime | None = None
    duration_ms: float = 0.0
    details: dict[str, Any] = field(default_factory=dict)
    error: str | None = None

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary."""
        return {
            "action_name": self.action_name,
            "action_type": self.action_type,
            "status": self.status.value,
            "message": self.message,
            "started_at": self.started_at.isoformat() if self.started_at else None,
            "completed_at": self.completed_at.isoformat() if self.completed_at else None,
            "duration_ms": self.duration_ms,
            "details": self.details,
            "error": self.error,
        }


@dataclass
class ActionContext:
    """Context passed to action execution.

    Contains all information needed for an action to execute,
    including the checkpoint result and environment variables.

    Attributes:
        checkpoint_result: The validation result.
        run_id: Unique run identifier.
        checkpoint_name: Name of the checkpoint.
        tags: Tags from the checkpoint.
        metadata: Additional metadata.
        environment: Environment variables (may contain secrets).
    """

    checkpoint_result: "CheckpointResult"
    run_id: str
    checkpoint_name: str
    tags: dict[str, str] = field(default_factory=dict)
    metadata: dict[str, Any] = field(default_factory=dict)
    environment: dict[str, str] = field(default_factory=dict)


@runtime_checkable
class ActionProtocol(Protocol):
    """Protocol for synchronous action implementations.

    Actions are executed after validation completes. Each action
    must implement the execute method.

    Example:
        class SlackNotification:
            def execute(self, context: ActionContext) -> ActionResult:
                # Send Slack message
                return ActionResult(
                    action_name="slack",
                    action_type="notification",
                    status=ActionStatus.SUCCESS,
                )
    """

    @property
    def name(self) -> str:
        """Get action name."""
        ...

    @property
    def action_type(self) -> str:
        """Get action type (notification, storage, webhook, custom)."""
        ...

    @property
    def config(self) -> ActionConfig:
        """Get action configuration."""
        ...

    def should_execute(self, context: ActionContext) -> bool:
        """Check if action should execute based on notify_on condition.

        Args:
            context: Execution context with checkpoint result.

        Returns:
            True if action should execute.
        """
        ...

    def execute(self, context: ActionContext) -> ActionResult:
        """Execute the action synchronously.

        Args:
            context: Execution context.

        Returns:
            Action result.
        """
        ...


@runtime_checkable
class AsyncActionProtocol(Protocol):
    """Protocol for asynchronous action implementations.

    For high-throughput scenarios, use async actions that don't
    block the event loop.

    Example:
        class AsyncSlackNotification:
            async def execute_async(self, context: ActionContext) -> ActionResult:
                async with aiohttp.ClientSession() as session:
                    await session.post(webhook_url, json=payload)
                return ActionResult(...)
    """

    @property
    def name(self) -> str:
        """Get action name."""
        ...

    @property
    def action_type(self) -> str:
        """Get action type."""
        ...

    @property
    def config(self) -> ActionConfig:
        """Get action configuration."""
        ...

    def should_execute(self, context: ActionContext) -> bool:
        """Check if action should execute."""
        ...

    async def execute_async(self, context: ActionContext) -> ActionResult:
        """Execute the action asynchronously.

        Args:
            context: Execution context.

        Returns:
            Action result.
        """
        ...


class CompensatableActionProtocol(Protocol):
    """Protocol for actions that support rollback (compensation).

    Compensatable actions can be rolled back if a subsequent action
    fails, supporting the Saga pattern for distributed transactions.

    Example:
        class DatabaseUpdateAction:
            def execute(self, context):
                self.backup_id = create_backup()
                update_database(context.checkpoint_result)
                return ActionResult(status=ActionStatus.SUCCESS)

            def compensate(self, context, execute_result):
                restore_from_backup(self.backup_id)
                return ActionResult(status=ActionStatus.COMPENSATED)
    """

    def execute(self, context: ActionContext) -> ActionResult:
        """Execute the forward action."""
        ...

    def compensate(
        self, context: ActionContext, execute_result: ActionResult
    ) -> ActionResult:
        """Compensate (rollback) the action.

        Args:
            context: Original execution context.
            execute_result: Result from the execute() call.

        Returns:
            Compensation result.
        """
        ...


class BaseAction(ABC):
    """Abstract base class for actions.

    Provides common functionality for all actions including
    notify_on condition checking and configuration management.

    Subclasses must implement:
    - action_type property
    - _do_execute method
    """

    def __init__(self, config: ActionConfig | dict[str, Any] | None = None) -> None:
        """Initialize action.

        Args:
            config: Action configuration or dict.
        """
        if config is None:
            self._config = ActionConfig()
        elif isinstance(config, dict):
            self._config = ActionConfig.from_dict(config)
        else:
            self._config = config

    @property
    def name(self) -> str:
        """Get action name."""
        return self._config.name or self.__class__.__name__

    @property
    @abstractmethod
    def action_type(self) -> str:
        """Get action type."""
        ...

    @property
    def config(self) -> ActionConfig:
        """Get action configuration."""
        return self._config

    def should_execute(self, context: ActionContext) -> bool:
        """Check if action should execute based on notify_on condition.

        Args:
            context: Execution context.

        Returns:
            True if action should execute.
        """
        if not self._config.enabled:
            return False

        result = context.checkpoint_result
        condition = self._config.notify_on

        if condition == NotifyCondition.ALWAYS:
            return True
        elif condition == NotifyCondition.SUCCESS:
            return result.status.value == "success"
        elif condition == NotifyCondition.FAILURE:
            return result.status.value == "failure"
        elif condition == NotifyCondition.ERROR:
            return result.status.value == "error"
        elif condition == NotifyCondition.FAILURE_OR_ERROR:
            return result.status.value in ("failure", "error")
        elif condition == NotifyCondition.NOT_SUCCESS:
            return result.status.value != "success"
        elif condition == NotifyCondition.WARNING:
            return result.status.value == "warning"

        return True

    def execute(self, context: ActionContext) -> ActionResult:
        """Execute the action.

        Args:
            context: Execution context.

        Returns:
            Action result.
        """
        if not self.should_execute(context):
            return ActionResult(
                action_name=self.name,
                action_type=self.action_type,
                status=ActionStatus.SKIPPED,
                message=f"Skipped due to notify_on={self._config.notify_on.value}",
            )

        started_at = datetime.now()
        try:
            result = self._do_execute(context)
            completed_at = datetime.now()
            result.started_at = started_at
            result.completed_at = completed_at
            result.duration_ms = (completed_at - started_at).total_seconds() * 1000
            return result
        except Exception as e:
            completed_at = datetime.now()
            return ActionResult(
                action_name=self.name,
                action_type=self.action_type,
                status=ActionStatus.FAILURE,
                message=f"Action failed: {str(e)}",
                started_at=started_at,
                completed_at=completed_at,
                duration_ms=(completed_at - started_at).total_seconds() * 1000,
                error=str(e),
            )

    @abstractmethod
    def _do_execute(self, context: ActionContext) -> ActionResult:
        """Perform the actual action execution.

        Subclasses must implement this method.

        Args:
            context: Execution context.

        Returns:
            Action result.
        """
        ...


class AsyncBaseAction(ABC):
    """Abstract base class for async actions.

    Similar to BaseAction but for asynchronous execution.
    """

    def __init__(self, config: ActionConfig | dict[str, Any] | None = None) -> None:
        """Initialize action."""
        if config is None:
            self._config = ActionConfig()
        elif isinstance(config, dict):
            self._config = ActionConfig.from_dict(config)
        else:
            self._config = config

    @property
    def name(self) -> str:
        """Get action name."""
        return self._config.name or self.__class__.__name__

    @property
    @abstractmethod
    def action_type(self) -> str:
        """Get action type."""
        ...

    @property
    def config(self) -> ActionConfig:
        """Get action configuration."""
        return self._config

    def should_execute(self, context: ActionContext) -> bool:
        """Check if action should execute."""
        if not self._config.enabled:
            return False

        result = context.checkpoint_result
        condition = self._config.notify_on

        if condition == NotifyCondition.ALWAYS:
            return True
        elif condition == NotifyCondition.SUCCESS:
            return result.status.value == "success"
        elif condition == NotifyCondition.FAILURE:
            return result.status.value == "failure"
        elif condition == NotifyCondition.ERROR:
            return result.status.value == "error"
        elif condition == NotifyCondition.FAILURE_OR_ERROR:
            return result.status.value in ("failure", "error")
        elif condition == NotifyCondition.NOT_SUCCESS:
            return result.status.value != "success"

        return True

    async def execute_async(self, context: ActionContext) -> ActionResult:
        """Execute the action asynchronously."""
        if not self.should_execute(context):
            return ActionResult(
                action_name=self.name,
                action_type=self.action_type,
                status=ActionStatus.SKIPPED,
                message=f"Skipped due to notify_on={self._config.notify_on.value}",
            )

        started_at = datetime.now()
        try:
            result = await self._do_execute_async(context)
            completed_at = datetime.now()
            result.started_at = started_at
            result.completed_at = completed_at
            result.duration_ms = (completed_at - started_at).total_seconds() * 1000
            return result
        except Exception as e:
            completed_at = datetime.now()
            return ActionResult(
                action_name=self.name,
                action_type=self.action_type,
                status=ActionStatus.FAILURE,
                message=f"Action failed: {str(e)}",
                started_at=started_at,
                completed_at=completed_at,
                duration_ms=(completed_at - started_at).total_seconds() * 1000,
                error=str(e),
            )

    @abstractmethod
    async def _do_execute_async(self, context: ActionContext) -> ActionResult:
        """Perform the actual async action execution."""
        ...


# =============================================================================
# Action Registry
# =============================================================================


class ActionRegistry:
    """Registry for action types.

    Allows registering and retrieving action implementations by name.
    Supports both built-in and custom action types.

    Example:
        registry = ActionRegistry()
        registry.register("slack", SlackNotificationAction)
        registry.register("email", EmailNotificationAction)

        action = registry.create("slack", config={"webhook_url": "..."})
    """

    def __init__(self) -> None:
        """Initialize registry."""
        self._actions: dict[str, type[BaseAction | AsyncBaseAction]] = {}
        self._factories: dict[str, Callable[..., BaseAction | AsyncBaseAction]] = {}

    def register(
        self,
        name: str,
        action_class: type[BaseAction | AsyncBaseAction],
    ) -> None:
        """Register an action class.

        Args:
            name: Action type name.
            action_class: Action class to register.
        """
        self._actions[name] = action_class

    def register_factory(
        self,
        name: str,
        factory: Callable[..., BaseAction | AsyncBaseAction],
    ) -> None:
        """Register an action factory function.

        Args:
            name: Action type name.
            factory: Factory function that creates actions.
        """
        self._factories[name] = factory

    def create(
        self,
        name: str,
        config: ActionConfig | dict[str, Any] | None = None,
        **kwargs: Any,
    ) -> BaseAction | AsyncBaseAction:
        """Create an action instance.

        Args:
            name: Action type name.
            config: Action configuration.
            **kwargs: Additional arguments for the action.

        Returns:
            Action instance.

        Raises:
            KeyError: If action type is not registered.
        """
        if name in self._factories:
            return self._factories[name](config=config, **kwargs)

        if name in self._actions:
            return self._actions[name](config=config, **kwargs)

        raise KeyError(f"Action type not found: {name}")

    def list_actions(self) -> list[str]:
        """List all registered action types.

        Returns:
            List of action type names.
        """
        return list(set(self._actions.keys()) | set(self._factories.keys()))

    def has_action(self, name: str) -> bool:
        """Check if an action type is registered.

        Args:
            name: Action type name.

        Returns:
            True if action type is registered.
        """
        return name in self._actions or name in self._factories


# Global action registry
_action_registry: ActionRegistry | None = None


def get_action_registry() -> ActionRegistry:
    """Get the global action registry.

    Returns:
        Global ActionRegistry instance.
    """
    global _action_registry
    if _action_registry is None:
        _action_registry = ActionRegistry()
    return _action_registry


def register_action(name: str) -> Callable[[type], type]:
    """Decorator to register an action class.

    Example:
        @register_action("my_custom")
        class MyCustomAction(BaseAction):
            ...
    """

    def decorator(cls: type) -> type:
        get_action_registry().register(name, cls)
        return cls

    return decorator
